ruby_programming/automated_testing/rspec_part_one_basics.md

```diff
@@ -0,0 +1,822 @@
+
+The `Gemfile.lock` was automatically generated after we ran the bundle install command earlier. It locks your gems down to specific versions so the project always uses the same versions no matter what machine the project is installed on.
+
+This is all you really need to know about Gemfile's for now, but we will include a link in the additional resources if you want to learn more about them.
+
+The lib directory is where our actual executable source code lives. In our case the todo list class.
+
+The spec directory is where we put all of our test files. We don't want to mix our test code with our source code as that would get messy very quickly and make it hard to discern the test code from the source code. Therefore it is best practice to put the test code in separate files within a different directory.
+
+Finally the `rspec --init` command we ran earlier generated a couple of files for us. The first of these was the  `.rspec` file, this is where we would add extra command line configuration for when we run the `rspec` command to execute our tests. Things like the format of the output RSpec produces in the terminal for example.
+
+The second file was the `spec_helper.rb` file, this is where we would add configuration that needs to be shared among many test files. We won't be doing anything more with either of these files during this lesson. But for completion sake, it's good to know what they are for.
+
+### Structure of a RSpec test file
+
+Open up the `todo_list_spec.rb` file we created earlier. The first thing we need to do is require a couple of dependencies. The first is the `spec_helper` file RSpec generated for us.
+
+The second file is the `todo_list.rb` file in lib so we can create an instance of the `TodoList` class in the our test file so we can invoke its methods and test what they do.
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+~~~
+
+#### Example Groups
+
+Now let's create the top level example group. Example groups serve two purposes, they are how we organize our test file by grouping related tests together and they allow us to *describe* the behaviour we are testing within them.
+
+We define an example group using the `describe` method. In the case of the following example group we are creating an example group for all the tests related to the `TodoList` Class.
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+end
+~~~
+
+Example groups can be nested within each other. This allows us to be more granular with grouping related tests together. We will see how that works by creating a nested example group where we can put all the tests for the `#name` method on the `TodoList` class.
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+	describe "name" do
```

Comment:
This is a minor nitpick, but I saw three places like this. This line does not visually render with the tab/spaces aligned correctly. It copied correctly to VS Code, but I think in some places it is a tab & other places it is two spaces.